pkg/kube/inject/webhook.go

```diff
@@ -1153,6 +1153,22 @@ func (wh *Webhook) inject(ar *kube.AdmissionReview, path string) *kube.Admission
 		} else {
 			log.Warnf("unable to fetch namespace, failed to get client for %q", clusterID)
 		}
+
+		// OpenShift automatically assigns a SecurityContext.RunAsUser to all containers in the Pod, even if the Pod's
+		// YAML does not explicitly set this value. Istio treats the values specified in the istio-proxy container as
+		// overrides and preserves them in the final Pod yaml as expected. However, the RunAsUser value which is
+		// automatically set by OpenShift would be the same for all containers within the Pod, which is a problem.
+		// Because the RunAsUser is identical for both the application container and the proxy container, traffic
+		// interception fails for the pod. Here, we ignore the RunAsUser value on the sidecar proxy if it matches the
+		// application container's value. At the same time, if user explicitly configures a RunAsUser in the istio-proxy
+		// container which is different to the application container's value, that setting is still honored.
+		if sideCarProxy := FindSidecar(params.pod); sideCarProxy != nil && sideCarProxy.SecurityContext != nil {
+			if isSidecarUserMatchingAppUser(params.pod.Spec.Containers) {
+				log.Infof("UserID of sideCar proxy %v matches with the app container, resetting", *sideCarProxy.SecurityContext.RunAsUser)
+				sideCarProxy.SecurityContext.RunAsUser = nil
```

Comment:
What happens once it is reset? How does it work after that point

Reply:
So as part of [injectPod](https://github.com/istio/istio/blob/994bed365bea44bb0727dc71dafbb1663efc4150/pkg/kube/inject/webhook.go#L454) when we return from RunTemplate() [here](https://github.com/istio/istio/blob/994bed365bea44bb0727dc71dafbb1663efc4150/pkg/kube/inject/webhook.go#L465), both the mergedPod/injectedPodData will have userId/groupId values derived from the annotation which is [present on the namespace](https://github.com/istio/istio/blob/994bed365bea44bb0727dc71dafbb1663efc4150/pkg/kube/inject/inject.go#L909) where the Pod was scheduled for OCP deployments. 

However, in the subsequent api (i.e., [reapplyOverwrittenContainers](https://github.com/istio/istio/blob/994bed365bea44bb0727dc71dafbb1663efc4150/pkg/kube/inject/webhook.go#L469)), we are overwriting the userIds based on the originalPod value to support [custom injection](https://istio.io/latest/docs/setup/additional-setup/sidecar-injection/#customizing-injection).

When the userIDs of sideCar proxy matches with application pod (i.e., when its automatically added by OCP), by resetting the value of the originalPod we ensure that reapplyOverwrittenContainers() method does not modify the userID of sideCar based on the originalPod value and the final mergedPod will have the value from the template.

I am adding the test-cases, hence marked this PR as draft.